sapicore/data/__init__.py

```diff
@@ -0,0 +1,474 @@
+            if destination.endswith(".csv"):
+                table = self.aggregate_descriptors()
+                pd.to_csv(table, ensure_dir(destination))
+
+        if kind is None or kind == "data":
+            if destination.endswith(".pt"):
+                torch.save(self.samples, ensure_dir(destination))
+
+            elif destination.endswith(".csv"):
+                pd.to_csv(self.samples, ensure_dir(destination))
+
+    def scan_root(self, pattern: str = "*", recursive: bool = False) -> list[str]:
+        """Scans the `root` directory and returns a list of files found that match the glob `pattern`
+        (similar to a regex).
+
+        Parameters
+        ----------
+        pattern: str, optional
+            Pattern to look for in file names that should be included (glob formatted).
+
+        recursive: bool, optional
+            Whether glob should search recursively. Defaults to `False`.
+
+        """
+        files = []
+        if os.path.exists(self.root):
+            for f in glob(os.path.join(self.root, pattern), recursive=recursive):
+                files.append(f)
+
+        return natsorted(files)
+
+    def select(self, conditions: str | list[str], axis: int = 0) -> list[int]:
+        """Selects sample indices by descriptor values based on a pandas logical statement applied to one or more
+        :class:`AxisDescriptor` objects, aggregated into a dataframe using :meth:`aggregate_descriptors`.
+
+        Importantly, these filtering operations can be performed prior to loading any data to memory, as they only
+        depend on :class:`AxisDescriptor` objects (labels) attached to this dataset.
+
+        Note
+        ----
+        Applying selection criteria with pd.eval syntax is simple:
+
+        * "age > 21" would execute `pd.eval("self.table.age > 21", target=df)`.
+        * "lobe.isin['F', 'P']" would execute `pd.eval("self.table.lobe.isin(['F', 'P']", target=df)`.
+
+        Where `df` is the table attribute of the :class:`~AxisDescriptor` the condition is applied to.
+
+        """
+        # wrap a single string condition in a list if necessary.
+        if isinstance(conditions, str):
+            conditions = [conditions]
+
+        # aggregate axis descriptors into a pandas dataframe (table).
+        df = self.aggregate_descriptors(axis=axis)
+
+        # evaluate the expressions and filter the dataframe.
+        for expression in conditions:
+            parsed = "& ".join([("df." + expr).replace(". ", ".") for expr in expression.split("&")])
+            parsed = "| ".join([("df." + expr).replace(". ", ".") for expr in parsed.split("|")])
+
+            parsed = parsed.replace("df.df.", "df.")
+            parsed = parsed.replace("df.(", "(df.")
+
+            df = df[pd.eval(parsed, target=df).to_list()]
+
+        # return indices where expressions evaluated to true.
+        return df.index.to_list()
+
+    def sample(self, method: Callable, axis: int = 0, **kwargs):
```

Comment:
I'm not quite sure how to evaluate the select/sample/trim methods because there are lots of detail that I'm not quite on board with. I can see why you added them, but I think they don't belong in the general `Data` class and I'd argue that it's trying to do too much when we should only be able to handle very simple data sets currently - those we actually have and not potential data sets.

I'd actually ask to leave these off for another PR if possible. Because e.g. I don't think we should be adding such general parsing as the select method does. To me, that's asking for complexity and too much magic and work for us down the line to support such things, when it may not be needed because I still think users should be responsible for writing simple data processing, rather than us creating generic things which have to be complex. And also I'm also not sure about using pandas in general as a backend for stuff.

Reply:
Safe to mark resolved after our discussion today?